jmclient/jmclient/wallet.py

```diff
@@ -1032,6 +999,501 @@ def close(self):
+                )) for k, v in psbt_output.derivation_map.items()}
+
+        if psbt_output.redeem_script:
+            outdict["redeem-script"] = bintohex(psbt_output.redeem_script)
+        if psbt_output.witness_script:
+            outdict["witness-script"] = bintohex(psbt_output.witness_script)
+
+        if psbt_output.proprietary_fields:
+            outdict["proprietary-fields"] = str(psbt_output.proprietary_fields)
+        if psbt_output.unknown_fields:
+            outdict["unknown-fields"] = str(psbt_output.unknown_fields)
+
+        return outdict
+
+    @staticmethod
+    def witness_utxos_to_psbt_utxos(utxos):
+        """ Given a dict of utxos as returned from select_utxos,
+        convert them to the format required to populate PSBT inputs,
+        namely CTxOut. Note that the non-segwit case is different, there
+        you should provide an entire CMutableTransaction object instead.
+        """
+        return [btc.CMutableTxOut(v["value"],
+                                  v["script"]) for _, v in utxos.items()]
+
+    def create_psbt_from_tx(self, tx, spent_outs=None):
+        """ Given a CMutableTransaction object, which should not currently
+        contain signatures, we create and return a new PSBT object of type
+        btc.PartiallySignedTransaction.
+        Optionally the information about the spent outputs that is stored
+        in PSBT_IN_NONWITNESS_UTXO, PSBT_IN_WITNESS_UTXO and PSBT_IN_REDEEM_SCRIPT
+        can also be provided, one item per input, in the tuple (spent_outs).
+        These objects should be either CMutableTransaction, CTxOut or None,
+        Note that redeem script information cannot be provided for inputs which
+        we don't own.
+        """
+        # TODO: verify tx contains no signatures as a sanity check?
+        new_psbt = btc.PartiallySignedTransaction(unsigned_tx=tx)
+        if spent_outs is None:
+            # user has not provided input script information; psbt
+            # will not yet be usable for signing.
+            return new_psbt
+        for i, txinput in enumerate(new_psbt.inputs):
+            if spent_outs[i] is None:
+                # as above, will not be signable in this case
+                continue
+            if isinstance(spent_outs[i], (btc.CTransaction, btc.CTxOut)):
+                # note that we trust the caller to choose Tx vs TxOut as according
+                # to non-witness/witness. Note also that for now this mixin does
+                # not attempt to provide previous tx (second argument) for witness
+                # case.
+                txinput.set_utxo(spent_outs[i], None)
+            else:
+                assert False, "invalid spent output type passed into PSBT creator"
+        # we now insert redeemscripts where that is possible and necessary:
+        for i, txinput in enumerate(new_psbt.inputs):
+            if isinstance(txinput.witness_utxo, btc.CTxOut):
```

Comment:
Note that you can also just do `if txinput.witness_utxo:` as `witness_utxo` is None if input is non-witness (and `txinput.nonwitness_utxo` will be None in the other case)

Reply:
quick question re wording: what is better/more correct English: `non_witness_utxo` or `nonwitness_utxo` ?

Reply:
Re: `if ...` : noted, I guess I will change that in next commit.

`nonwitness_utxo` feels more natural to me, though I doubt there is an objective answer.

Reply:
Actually I won't bother to change this, the check is more stringent than it needs to be, but it's fine.

Reply:
Agreed